### PR TITLE
[FLINK-19605][table-runtime-blink] Implement cumulative windowing for window aggregate operator

### PR DESCRIPTION
**Pull request by wuchong**
## What is the purpose of the change

Implements the cumulative window assigner and supports cumulative window aggregate in `WindowOperator`.

## Brief change log

- Add `CumulativeWindowAssigner`
- Add cumulative window builder method in `WindowOperatorBuilder`

## Verifying this change

- Add `CumulativeWindowAssignerTest` to test `CumulativeWindowAssigner`.
- Add harness tests for cumulative window aggregate in `WindowOperatorTest`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by wuchong** [comment]
cc @danny0405

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/window/assigners/CumulativeWindowAssigner.java [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+			throw new IllegalArgumentException(
+				"CumulativeWindowAssigner parameters must satisfy step > 0 and size > 0");
+		}
+		if (size % step != 0) {
+			throw new IllegalArgumentException(
+				"CumulativeWindowAssigner requires size must be an integral multiple of step.");
+		}
+
+		this.size = size;
+		this.step = step;
```
[the remark this answers is not in the document]
Comment (wuchong) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/window/assigners/CumulativeWindowAssigner.java:62]: Yes. We can support `2 days` max size.

**Comment by wuchong** [comment]
@danny0405 please have a look again.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/window/assigners/CumulativeWindowAssigner.java [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+	/**
+	 * Creates a new {@link CumulativeWindowAssigner} that assigns
+	 * elements to cumulative time windows based on the element timestamp.
+	 *
+	 * @param size  The max size of the generated windows.
+	 * @param step The step interval for window size to increase of the generated windows.
+	 * @return The time policy.
+	 */
+	public static CumulativeWindowAssigner of(Duration size, Duration step) {
+		return new CumulativeWindowAssigner(size.toMillis(), step.toMillis(), 0, true);
```
Comment (danny0405) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/window/assigners/CumulativeWindowAssigner.java:156]: `size` -> `maxSize`

**Review by danny0405 — APPROVED** [review]

**Comment by wuchong** [comment]
The failed Azure is the unstable e2e tests. This change shouldn't affect e2e part. 

Merging...

**Comment by HuangZhenQiu** [comment]
@danny0405 @wuchong 
As I remember, there was a FLIP talking about the cumulative window. But I can't find it right now. Would you please point me to the design doc? We have some similar requirements internally, so want to align more on the functionality with you guys.

**Comment by wuchong** [comment]
@HuangZhenQiu , here is the link https://cwiki.apache.org/confluence/display/FLINK/FLIP-145%3A+Support+SQL+windowing+table-valued+function

**Pull request merged**